docs/architecture/adr-039-epoched-staking.md

```diff
@@ -0,0 +1,110 @@
+
+### Staking messages
+- **MsgCreateValidator**: Move user's self-bond to `EpochDelegationPool` immediately. Queue a message for the epoch boundary to handle the self-bond, taking the funds from the `EpochDelegationPool`. If Epoch execution fail, return back funds from `EpochDelegationPool` to user's account.
+- **MsgEditValidator**: Validate message and if valid queue the message for execution at the end of the Epoch.
+- **MsgDelegate**: Move user's funds to `EpochDelegationPool` immediately. Queue a message for the epoch boundary to handle the delegation, taking the funds from the `EpochDelegationPool`. If Epoch execution fail, return back funds from `EpochDelegationPool` to user's account.
+- **MsgBeginRedelegate**: Validate message and if valid queue the message for execution at the end of the Epoch.
+- **MsgUndelegate**: Validate message and if valid queue the message for execution at the end of the Epoch.
+
+### Slashing messages
+- **MsgUnjail**: Validate message and if valid queue the message for execution at the end of the Epoch.
+- **Slash Event**: Whenever a slash event is created, it gets queued in the slashing module to apply at the end of the epoch. The queues should be setup such that this slash applies immediately.
+
+### Evidence Messages
+ - **MsgSubmitEvidence**: This gets executed immediately, and the validator gets jailed immediately. However in slashing, the actual slash event gets queued.
+
+Then we add methods to the end blockers, to ensure that at the epoch boundary the queues are cleared and delegation updates are applied.
+
+
+__Step-2__: Implement querying of queued staking txs. 
+
+When querying the staking activity of a given address, the status should return not only the amount of tokens staked, but also if there are any queued stake events for that address. This will require nodes supporting querying to either do some more indexing to have this be efficiently queryable, or to have transactions 
+
+__Step-3__: Adjust gas
+
+Currently gas represents the cost of executing a transaction when its done immediately. (Merging together costs of p2p overhead, state access overhead, and computational overhead) However, now a transaction can cause computation in a future block, namely at the epoch boundary.
+
+To handle this, we should initially include parameters for estimating the amount of future computation (denominated in gas), and add that as a flat charge needed for the message.
+We leave it as out of scope for how to weight future computation versus current computation in gas pricing, and have it set such that the are weighted equally for now.
+
+## Consequences
+
+### Positive
+
+* Abstracts the proof of stake module that allows retaining the existing functionality
+* Enables new features such as validator-set based threshold cryptography
+
+### Negative
+
+* Increases complexity of integrating more complex gas pricing mechanisms, as they now have to consider future execution costs as well.
```

Comment:
* When epoch variable is > 1 validators may no longer exit the system arbitrarily, but at an epoch.